keys.go

```diff
@@ -0,0 +1,174 @@
+	}
+	var keys ServerKeys
+	if keys.Raw, err = ioutil.ReadAll(response.Body); err != nil {
+		return nil, nil, err
+	}
+	if err = json.Unmarshal(keys.Raw, &keys); err != nil {
+		return nil, nil, err
+	}
+	return &keys, &connectionState, nil
+}
+
+// KeyChecks are the checks that should be applied to ServerKey responses.
+type KeyChecks struct {
+	AllChecksOK               bool // Did all the checks pass?
+	MatchingServerName        bool // Does the server name match what was requested.
+	FutureValidUntilTS        bool // The valid until TS is in the future.
+	HasEd25519VerifyKey       bool // The key response has a E25519 key for the server.
+	ValidEd25519VerifyKey     bool // The verify keys are value.
+	MatchingEd25519Signature  bool // Every verify key claimed has a valid signature.
+	HasTLSFingerprint         bool // The response includes a TLS fingerprint.
+	ValidSHA256TLSFingerprint bool // Every TLS fingerprint includes a SHA-256 hash.
+	MatchingTLSFingerprint    bool // The TLS fingerprint for the connection matches one of the listed fingerprints.
+}
+
+// CheckKeys checks the keys returned from a server to make sure they are valid.
+// If the checks pass the also returns a map of key_id to Ed25519 public key and a list of SHA256 TLS fingerprints.
+func CheckKeys(serverName string, timeNowMs uint64, keys ServerKeys, connState *tls.ConnectionState) (
+	checks KeyChecks, ed25519Keys map[string]Base64String, sha256Fingerprints []Base64String,
+) {
+	checks.MatchingServerName = serverName == keys.ServerName
+
+	ed25519Keys = checkVerifyKeys(keys, &checks)
+	sha256Fingerprints = checkTLSFingerprints(keys, &checks)
+
+	checks.FutureValidUntilTS = timeNowMs < keys.ValidUntilTS
+
+	if connState == nil {
+		// No TLS connection state so skip the fingerprint checks.
+		checks.MatchingTLSFingerprint = true
+	} else {
+		// Check the peer certificates.
+		checks.MatchingTLSFingerprint = checkFingerprint(connState, sha256Fingerprints)
+	}
+
+	checks.AllChecksOK = (checks.MatchingServerName &&
+		checks.FutureValidUntilTS &&
+		checks.HasEd25519VerifyKey &&
+		checks.ValidEd25519VerifyKey &&
+		checks.MatchingEd25519Signature &&
+		checks.HasTLSFingerprint &&
+		checks.ValidSHA256TLSFingerprint &&
+		checks.MatchingTLSFingerprint)
+
+	if !checks.AllChecksOK {
+		sha256Fingerprints = nil
+		ed25519Keys = nil
+	}
+	return
+}
+
+func checkFingerprint(connState *tls.ConnectionState, sha256Fingerprints []Base64String) bool {
+	if len(connState.PeerCertificates) == 0 {
+		return false
+	}
+	cert := connState.PeerCertificates[0]
+	digest := sha256.Sum256(cert.Raw)
+	for _, fingerprint := range sha256Fingerprints {
+		if bytes.Compare(digest[:], fingerprint) == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func checkVerifyKeys(keys ServerKeys, checks *KeyChecks) map[string]Base64String {
+	checks.ValidEd25519VerifyKey = true
+	checks.MatchingEd25519Signature = true
+	verifyKeys := map[string]Base64String{}
+	for keyID, keyData := range keys.VerifyKeys {
```

Comment:
I really dislike not failing safely. If `len(keys.VerifyKeys) == 0` then both `ValidEd25519VerifyKey` and `MatchingEd25519Signature` will be true.